content/api/servo-assembly/arduino-commands.md

```diff
@@ -0,0 +1,196 @@
+than can be achieved through the Python API.
+
+## Requirements
+
+As your new firmware will still need to be able to talk to the Python code
+running on the Raspberry Pi, you'll need to base it on SourceBots default
+firmware. You can get a copy of the source code from GitHub either by
+downloading the [repo archive](servo-firmware-zip) or by cloning the repo:
+
+```txt
+git clone https://github.com/sourcebots/servo-firmware
+```
+
+We recommend using [Arduino Studio](https://www.arduino.cc/en/Main/Software) to
+edit your firmware as it provides direct support for building the code into the
+firmware binary and downloading it onto the Arduino.
+
+## Adding a command
+
+In order for the code running on your Pi to be able to interact with the
+functionality you're adding to the Arduino, you'll need to define a new command.
+
+Adding a command requires two changes. First you must define the function which
+will implement the operations of that command and then you register that command
+under the name you'll use to call it from the Python code.
+
+### Implementing the command
+
+Here's a simple command which just echoes back any arguments it receives:
+
+```cpp
+// Define a function which implements your command. As long as you stick to
+// letters, numbers and underscore (_), you can call it whatever you like.
+// With the exception of the name (here it's `echo`), the rest of the function's
+// "signature" must be exactly as it is on the following line:
+static CommandError echo(int commandId, String argument) {
+
+  if (argument.length() == 0) {
+    // If something went wrong, you should return a `COMMAND_ERROR` with a
+    // message about the failure. That message will be fed back to Python where
+    // an exception will be raised.
+    return COMMAND_ERROR("nothing to echo");
+  }
+
+  while (argument.length() > 0) {
+    // You can use `pop_option` to extract individual parameters from the
+    // `argument` string. They're sent space-separated, so be sure not to
+    // include any spaces within your argument values!
+    //
+    // You don't need to check the length of `argument` before calling
+    // `pop_option` -- if there isn't an option to be had you'll get back an
+    // empty string ("").
+    String next_arg = pop_option(argument);
+
+    // You can write values to be read by the code on the Pi using the
+    // `serialWrite` function. The first two arguments need to be exactly as
+    // they are here; the third argument needs to be a `String` containing the
+    // value you want to send back to the Pi.
+    serialWrite(commandId, '>', next_arg);
+  }
+
+  // If everything went fine, you should return `OK`.
+  return OK;
+}
+```
+<!-- TODO: mention being able to send back comments? -->
+
+For other examples of how to define commands (and the helper functions you can
+use), you can have a look at the other commands the Arduino already supports.
+In particular the `get_version`, `led` and `servo`  are a good starting point.
```

Comment:
Are you missing the word "commands" after "servo" here?

Reply:
almost: 932d341